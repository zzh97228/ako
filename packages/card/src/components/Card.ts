import vue, { VNodeArrayChildren } from 'vue';
import { defineComponent, h, mergeProps, computed } from 'vue';
import { useSize, sizeProps, useColor, useElevation, genElevationProp, genColorProp } from '@lagabu/shared';

export default defineComponent({
  name: 'card',
  props: {
    tag: {
      type: String,
      default: 'div',
    },
    outlined: Boolean,
    flat: Boolean,
    tile: Boolean,
    ...genElevationProp(),
    ...sizeProps,
    ...genColorProp(),
  },
  setup(props, { slots }) {
    const { sizeStyle } = useSize(props);
    const elevation = useElevation(props);
    const isOutlined = computed(() => props.outlined);
    const color = useColor(props, isOutlined);
    const classes = computed(() => {
      return {
        card: true,
        'card--outlined': props.outlined,
        'card--flat': props.flat,
        'card--tile': props.tile,
      };
    });

    const genImage = (children?: VNodeArrayChildren) => {
      return h(
        'div',
        {
          class: 'card__image',
        },
        children
      );
    };

    return () =>
      h(
        props.tag,
        mergeProps(
          {
            class: classes.value,
            style: sizeStyle.value,
          },
          {
            class: color.class.value,
            style: color.style.value,
          },
          {
            class: elevation.class.value,
            style: elevation.style.value,
          }
        ),
        [
          slots.image && genImage(slots.image()),
          h(
            'div',
            {
              class: 'card__content',
            },
            slots.default && slots.default()
          ),
        ]
      );
  },
});
