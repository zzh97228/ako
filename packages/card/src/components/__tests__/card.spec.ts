import { mount, VueWrapper } from '@vue/test-utils';
import Card from '../Card';

describe('card.ts', () => {
  let mountFunc: (options?: object) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (opts = {}) => mount(Card, opts);
  });
  it('should render card', () => {
    const wrapper = mountFunc({
      props: {
        width: '100%',
        height: '100%',
      },
    });

    expect(wrapper.classes()).toContain('card');
    expect(wrapper.classes()).toContain('elevation-xs');
    expect(wrapper.attributes('style')).toContain('width: 100%');
    expect(wrapper.attributes('style')).toContain('height: 100%');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
