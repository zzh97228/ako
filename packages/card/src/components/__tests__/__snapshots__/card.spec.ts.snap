// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card.ts should render card 1`] = `
<div class="card card--outlined elevation-xs"
     style="width: 100%; height: 100%;"
>
  <div class="card__content">
  </div>
</div>
`;
